src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java

```diff
@@ -394,5 +398,70 @@ public CloudSdk build() {
           startListeners, runDevAppServerWaitSeconds);
     }
 
+    /**
+     * Attempt to find the Google Cloud SDK in various places.
+     * 
+     * @return the path to the root of the Google Cloud SDK
+     * @throws AppEngineException if not found
+     */
+    @Nonnull
+    private Path discoverSdkPath() {
+      for (CloudSdkResolver resolver : getResolvers()) {
+        try {
+          Path discoveredSdkPath = resolver.getCloudSdkPath();
+          if (discoveredSdkPath != null) {
+            return discoveredSdkPath;
+          }
+        } catch (RuntimeException ex) {
+          // prevent interference from exceptions in other resolvers
+          logger.log(Level.SEVERE, resolver.getClass().getName()
+              + ": exception thrown when searching for Google Cloud SDK", ex);
+        }
+      }
+      throw new AppEngineException("Google Cloud SDK path was not provided and could not be"
+          + " found in any known install locations.");
+    }
+
+    /**
+     * Return the configured SDK resolvers.
+     */
+    @VisibleForTesting
+    public List<CloudSdkResolver> getResolvers() {
+      List<CloudSdkResolver> resolvers;
+      if (this.resolvers != null) {
+        resolvers = new ArrayList<>(this.resolvers);
+      } else {
+        // Explicitly specify classloader rather than use the Thread Context Class Loader (TCCL)
+        ServiceLoader<CloudSdkResolver> services =
+            ServiceLoader.load(CloudSdkResolver.class, getClass().getClassLoader());
+        resolvers = Lists.newArrayList(services);
+        // Explicitly add the PATH-based resolver
+        resolvers.add(new PathResolver());
+      }
+      Collections.sort(resolvers, new ResolverComparator());
+      return resolvers;
+    }
+
+    /*
+     * Set the list of path resolvers to locate the Google Cloud SDK. Intended for tests to
+     * precisely control where the SDK may be found.
+     */
+    @VisibleForTesting
+    public Builder resolvers(List<CloudSdkResolver> resolvers) {
+      this.resolvers = resolvers;
+      return this;
+    }
+  }
+
+  /**
+   * Compare two {@link CloudSdkResolver} instances by their rank.
+   */
+  private static class ResolverComparator implements Comparator<CloudSdkResolver>, Serializable {
```

Comment:
we should remove that check from our FindBugs config. It's not relevant to us. 

http://findbugs.sourceforge.net/bugDescriptions.html#SE_COMPARATOR_SHOULD_BE_SERIALIZABLE